staging/src/kubevirt.io/client-go/apis/flavor/v1alpha1/types.go

```diff
@@ -0,0 +1,93 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1 "kubevirt.io/client-go/api/v1"
+)
+
+// VirtualMachineFlavor resource contains common VirtualMachine configuration
+// that can be used by multiple VirtualMachine resources.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +genclient
+type VirtualMachineFlavor struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +listType=map
+	// +listMapKey=name
+	Profiles []VirtualMachineFlavorProfile `json:"profiles"`
+}
+
+// VirtualMachineFlavorList is a list of VirtualMachineFlavor resources.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+type VirtualMachineFlavorList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []VirtualMachineFlavor `json:"items"`
+}
+
+// VirtualMachineClusterFlavor is a cluster scoped version of VirtualMachineFlavor resource.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +genclient
+// +genclient:nonNamespaced
+type VirtualMachineClusterFlavor struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +listType=map
+	// +listMapKey=name
+	Profiles []VirtualMachineFlavorProfile `json:"profiles"`
+}
+
+// VirtualMachineClusterFlavorList is a list of VirtualMachineClusterFlavor resources.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+type VirtualMachineClusterFlavorList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []VirtualMachineClusterFlavor `json:"items"`
+}
+
+// VirtualMachineFlavorProfile contains definitions that will be applied to VirtualMachine.
+//
+// +k8s:openapi-gen=true
+type VirtualMachineFlavorProfile struct {
+	// Name specifies the name of this custom profile.
+	Name string `json:"name"`
+
+	// Default specifies if this VirtualMachineFlavorProfile is the default for the VirtualMachineFlavor.
+	// Zero or one profile can be set to default.
+	//
+	// +optional
+	Default bool `json:"default,omitempty"`
```

Comment:
When I was reading the design document, I did not notice the idea of **layered** custom profiles. There was no mention of merging the default and custom profiles, and I understood it as custom profiles being unrelated to the default profile.
Thank you for pointing this out, now it makes more sense.

Now that I know this, I would prefer the the layered + default approach, because it will be easier for the user.
Most of the merge conflicts can be solved by preferring the value in custom profile. Problematic case will be merging arrays. I cannot tell now if we can work around it or not, It depends on the specific fields in the flavor.

Reply:
> Problematic case will be merging arrays. I cannot tell now if we can work around it or not, It depends on the specific fields in the flavor

yup, and that's starting to be a bit of my concern. We had to deal with this sort of merging when we created VMPresets, and it makes for some awkward scenarios.

your example of how to handle lists illustrates this well. Do we take the custom profiles list, or merge the two lists together? and if we merge the two lists together then how could we disable/remove an entry in the default profile by using a custom profile if we needed to?

I'm not completely sold on my original intent behind layering.

Reply:
for the layering idea, i wouldn't call it default but base profile. that makes it clearer that it's layered on top.
If it's called default, I would expect an either-or result and no combination of profiles.

Reply:
And I think I mentioned this before, if you want to use the same merging logic Server-Side apply uses, have a look at https://github.com/kubernetes-sigs/structured-merge-diff
It's a bit complicated, but usable and could give you a lot if you want proper schema based merging.

Reply:
I think we can define the profile `struct` in a way that the merge operation would be unambiguous. So confusing scenarios would not happen. For example, we could use maps instead of lists, or use lists with some "key" field and merge them as maps.